File: SocketClient.cpp
/*
 * DBK驱动Socket通信客户端 (C++版本)
 * 用于通过Socket与内核驱动进行通信，替代传统的IRP/DeviceIoControl方式
 */

#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>
#include <stdio.h>
#include <stdint.h>

#pragma comment(lib, "ws2_32.lib")

// Socket配置
#define SERVER_HOST "127.0.0.1"
#define SERVER_PORT 28996

// IOCTL代码定义
#define IOCTL_CE_READMEMORY         0x9C402000
#define IOCTL_CE_WRITEMEMORY        0x9C402004
#define IOCTL_CE_OPENPROCESS        0x9C402008
#define IOCTL_CE_GETVERSION         0x9C4020C0
#define IOCTL_CE_GETPEPROCESS       0x9C402034
#define IOCTL_CE_READPHYSICALMEMORY 0x9C402038

// 消息头结构
#pragma pack(push, 1)
typedef struct _SOCKET_MESSAGE_HEADER {
    ULONG IoControlCode;
    ULONG InputBufferSize;
    ULONG OutputBufferSize;
    ULONG Reserved;
} SOCKET_MESSAGE_HEADER, *PSOCKET_MESSAGE_HEADER;

typedef struct _SOCKET_RESPONSE_HEADER {
    LONG Status;
    ULONG DataSize;
    ULONG Reserved1;
    ULONG Reserved2;
} SOCKET_RESPONSE_HEADER, *PSOCKET_RESPONSE_HEADER;
#pragma pack(pop)

class DBKSocketClient {
private:
    SOCKET sock;
    bool connected;

    bool RecvExact(void* buffer, size_t size) {
        char* ptr = (char*)buffer;
        size_t received = 0;
        
        while (received < size) {
            int ret = recv(sock, ptr + received, (int)(size - received), 0);
            if (ret <= 0) {
                printf("[-] 接收数据失败: %d\n", WSAGetLastError());
                return false;
            }
            received += ret;
        }
        return true;
    }

public:
    DBKSocketClient() : sock(INVALID_SOCKET), connected(false) {
        WSADATA wsaData;
        WSAStartup(MAKEWORD(2, 2), &wsaData);
    }

    ~DBKSocketClient() {
        Disconnect();
        WSACleanup();
    }

    bool Connect(const char* host = SERVER_HOST, int port = SERVER_PORT) {
        sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
        if (sock == INVALID_SOCKET) {
            printf("[-] 创建Socket失败: %d\n", WSAGetLastError());
            return false;
        }

        sockaddr_in serverAddr;
        serverAddr.sin_family = AF_INET;
        serverAddr.sin_port = htons(port);
        inet_pton(AF_INET, host, &serverAddr.sin_addr);

        if (connect(sock, (sockaddr*)&serverAddr, sizeof(serverAddr)) == SOCKET_ERROR) {
            printf("[-] 连接失败: %d\n", WSAGetLastError());
            closesocket(sock);
            sock = INVALID_SOCKET;
            return false;
        }

        connected = true;
        printf("[+] 成功连接到驱动 %s:%d\n", host, port);
        return true;
    }

    void Disconnect() {
        if (sock != INVALID_SOCKET) {
            closesocket(sock);
            sock = INVALID_SOCKET;
            connected = false;
            printf("[+] 已断开连接\n");
        }
    }

    bool SendRequest(ULONG ioctlCode, void* inputData, ULONG inputSize, 
                     void* outputData, ULONG outputSize, LONG* status) {
        if (!connected) {
            printf("[-] 未连接到服务器\n");
            return false;
        }

        // 发送消息头
        SOCKET_MESSAGE_HEADER msgHeader;
        msgHeader.IoControlCode = ioctlCode;
        msgHeader.InputBufferSize = inputSize;
        msgHeader.OutputBufferSize = outputSize;
        msgHeader.Reserved = 0;

        if (send(sock, (char*)&msgHeader, sizeof(msgHeader), 0) == SOCKET_ERROR) {
            printf("[-] 发送消息头失败: %d\n", WSAGetLastError());
            return false;
        }

        // 发送输入数据
        if (inputSize > 0 && inputData) {
            if (send(sock, (char*)inputData, inputSize, 0) == SOCKET_ERROR) {
                printf("[-] 发送输入数据失败: %d\n", WSAGetLastError());
                return false;
            }
        }

        // 接收响应头
        SOCKET_RESPONSE_HEADER respHeader;
        if (!RecvExact(&respHeader, sizeof(respHeader))) {
            return false;
        }

        if (status) {
            *status = respHeader.Status;
        }

        // 接收响应数据
        if (respHeader.DataSize > 0 && outputData) {
            if (!RecvExact(outputData, min(respHeader.DataSize, outputSize))) {
                return false;
            }
        }

        return true;
    }

    bool GetVersion(ULONG* version) {
        LONG status;
        if (!SendRequest(IOCTL_CE_GETVERSION, NULL, 0, version, sizeof(ULONG), &status)) {
            return false;
        }

        if (status == 0) {
            printf("[+] 驱动版本: %u\n", *version);
            return true;
        } else {
            printf("[-] 获取版本失败，状态码: 0x%08X\n", status);
            return false;
        }
    }

    bool OpenProcess(DWORD processId, UINT64* handle) {
        struct {
            UINT64 h;
            BYTE special;
        } output;

        LONG status;
        if (!SendRequest(IOCTL_CE_OPENPROCESS, &processId, sizeof(DWORD), 
                        &output, sizeof(output), &status)) {
            return false;
        }

        if (status == 0) {
            *handle = output.h;
            printf("[+] 打开进程 PID=%u 成功\n", processId);
            printf("    句柄: 0x%016llX\n", output.h);
            printf("    特殊标志: %d\n", output.special);
            return true;
        } else {
            printf("[-] 打开进程失败，状态码: 0x%08X\n", status);
            return false;
        }
    }

    bool ReadProcessMemory(UINT64 processId, UINT64 address, WORD size, void* buffer) {
        struct {
            UINT64 processid;
            UINT64 startaddress;
            WORD bytestoread;
        } input;

        input.processid = processId;
        input.startaddress = address;
        input.bytestoread = size;

        BYTE* tempBuffer = new BYTE[size + sizeof(input)];
        LONG status;

        if (!SendRequest(IOCTL_CE_READMEMORY, &input, sizeof(input), 
                        tempBuffer, size + sizeof(input), &status)) {
            delete[] tempBuffer;
            return false;
        }

        if (status == 0) {
            memcpy(buffer, tempBuffer + sizeof(input), size);
            printf("[+] 读取内存成功: PID=%llu, 地址=0x%llX, 大小=%u\n", 
                   processId, address, size);
            delete[] tempBuffer;
            return true;
        } else {
            printf("[-] 读取内存失败，状态码: 0x%08X\n", status);
            delete[] tempBuffer;
            return false;
        }
    }

    bool GetPEProcess(DWORD processId, UINT64* peprocess) {
        LONG status;
        if (!SendRequest(IOCTL_CE_GETPEPROCESS, &processId, sizeof(DWORD), 
                        peprocess, sizeof(UINT64), &status)) {
            return false;
        }

        if (status == 0) {
            printf("[+] EPROCESS地址: 0x%016llX\n", *peprocess);
            return true;
        } else {
            printf("[-] 获取EPROCESS失败，状态码: 0x%08X\n", status);
            return false;
        }
    }
};

void PrintHexDump(const void* data, size_t size, UINT64 baseAddress = 0) {
    const BYTE* bytes = (const BYTE*)data;
    for (size_t i = 0; i < size; i += 16) {
        printf("  %08llX: ", baseAddress + i);
        
        // 打印十六进制
        for (size_t j = 0; j < 16; j++) {
            if (i + j < size) {
                printf("%02X ", bytes[i + j]);
            } else {
                printf("   ");
            }
        }
        
        printf(" ");
        
        // 打印ASCII
        for (size_t j = 0; j < 16 && i + j < size; j++) {
            BYTE b = bytes[i + j];
            printf("%c", (b >= 32 && b < 127) ? b : '.');
        }
        
        printf("\n");
    }
}

int main() {
    printf("============================================================\n");
    printf("DBK驱动Socket通信客户端测试 (C++版本)\n");
    printf("============================================================\n");

    DBKSocketClient client;

    // 连接到驱动
    if (!client.Connect()) {
        return 1;
    }

    // 测试1: 获取驱动版本
    printf("\n[测试1] 获取驱动版本\n");
    printf("------------------------------------------------------------\n");
    ULONG version;
    client.GetVersion(&version);

    // 测试2: 打开当前进程
    printf("\n[测试2] 打开当前进程\n");
    printf("------------------------------------------------------------\n");
    DWORD currentPid = GetCurrentProcessId();
    printf("当前进程PID: %u\n", currentPid);
    UINT64 handle;
    client.OpenProcess(currentPid, &handle);

    // 测试3: 获取EPROCESS
    printf("\n[测试3] 获取EPROCESS结构\n");
    printf("------------------------------------------------------------\n");
    UINT64 peprocess;
    client.GetPEProcess(currentPid, &peprocess);

    // 测试4: 读取内存
    printf("\n[测试4] 读取进程内存\n");
    printf("------------------------------------------------------------\n");
    UINT64 testAddress = 0x400000; // 典型的PE文件基址
    BYTE memory[64];
    if (client.ReadProcessMemory(currentPid, testAddress, sizeof(memory), memory)) {
        printf("内存数据:\n");
        PrintHexDump(memory, sizeof(memory), testAddress);
    }

    printf("\n============================================================\n");
    printf("测试完成！\n");
    printf("============================================================\n");

    client.Disconnect();
    
    printf("\n按任意键退出...");
    getchar();
    return 0;
}

File: SocketClient.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
DBK驱动Socket通信客户端
用于通过Socket与内核驱动进行通信，替代传统的IRP/DeviceIoControl方式
"""

import socket
import struct
import sys

# Socket配置
SERVER_HOST = "127.0.0.1"
SERVER_PORT = 28996

# IOCTL代码定义（从原驱动中提取）
IOCTL_CE_READMEMORY = 0x9C402000
IOCTL_CE_WRITEMEMORY = 0x9C402004
IOCTL_CE_OPENPROCESS = 0x9C402008
IOCTL_CE_GETVERSION = 0x9C4020C0
IOCTL_CE_GETPEPROCESS = 0x9C402034
IOCTL_CE_READPHYSICALMEMORY = 0x9C402038

class DBKSocketClient:
    """DBK驱动Socket客户端"""
    
    def __init__(self, host=SERVER_HOST, port=SERVER_PORT):
        self.host = host
        self.port = port
        self.sock = None
        
    def connect(self):
        """连接到驱动Socket服务器"""
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.connect((self.host, self.port))
            print(f"[+] 成功连接到驱动 {self.host}:{self.port}")
            return True
        except Exception as e:
            print(f"[-] 连接失败: {e}")
            return False
    
    def disconnect(self):
        """断开连接"""
        if self.sock:
            self.sock.close()
            self.sock = None
            print("[+] 已断开连接")
    
    def send_request(self, ioctl_code, input_data=b"", output_size=0):
        """
        发送请求到驱动
        
        参数:
            ioctl_code: IOCTL控制码
            input_data: 输入数据（bytes）
            output_size: 期望的输出数据大小
            
        返回:
            (status, output_data) 元组
        """
        if not self.sock:
            print("[-] 未连接到服务器")
            return None, None
        
        try:
            # 构造消息头
            # struct SOCKET_MESSAGE_HEADER {
            #     ULONG IoControlCode;
            #     ULONG InputBufferSize;
            #     ULONG OutputBufferSize;
            #     ULONG Reserved;
            # }
            input_size = len(input_data)
            header = struct.pack("<IIII", ioctl_code, input_size, output_size, 0)
            
            # 发送消息头
            self.sock.sendall(header)
            
            # 发送输入数据
            if input_size > 0:
                self.sock.sendall(input_data)
            
            # 接收响应头
            # struct SOCKET_RESPONSE_HEADER {
            #     NTSTATUS Status;
            #     ULONG DataSize;
            #     ULONG Reserved1;
            #     ULONG Reserved2;
            # }
            resp_header = self._recv_exact(16)
            if not resp_header:
                print("[-] 接收响应头失败")
                return None, None
            
            status, data_size, _, _ = struct.unpack("<iIII", resp_header)
            
            # 接收响应数据
            output_data = b""
            if data_size > 0:
                output_data = self._recv_exact(data_size)
                if not output_data:
                    print("[-] 接收响应数据失败")
                    return status, None
            
            return status, output_data
            
        except Exception as e:
            print(f"[-] 发送请求失败: {e}")
            return None, None
    
    def _recv_exact(self, size):
        """接收指定大小的数据"""
        data = b""
        while len(data) < size:
            chunk = self.sock.recv(size - len(data))
            if not chunk:
                return None
            data += chunk
        return data
    
    def get_version(self):
        """获取驱动版本"""
        status, data = self.send_request(IOCTL_CE_GETVERSION, b"", 4)
        if status == 0 and data:
            version = struct.unpack("<I", data)[0]
            print(f"[+] 驱动版本: {version}")
            return version
        else:
            print(f"[-] 获取版本失败，状态码: 0x{status:08X}")
            return None
    
    def open_process(self, process_id):
        """打开进程"""
        input_data = struct.pack("<I", process_id)
        status, data = self.send_request(IOCTL_CE_OPENPROCESS, input_data, 9)
        
        if status == 0 and data:
            handle, special = struct.unpack("<QB", data)
            print(f"[+] 打开进程 PID={process_id} 成功")
            print(f"    句柄: 0x{handle:016X}")
            print(f"    特殊标志: {special}")
            return handle
        else:
            print(f"[-] 打开进程失败，状态码: 0x{status:08X}")
            return None
    
    def read_process_memory(self, process_id, address, size):
        """读取进程内存"""
        # struct input {
        #     UINT64 processid;
        #     UINT64 startaddress;
        #     WORD bytestoread;
        # }
        input_data = struct.pack("<QQH", process_id, address, size)
        status, data = self.send_request(IOCTL_CE_READMEMORY, input_data, size + 18)
        
        if status == 0 and data:
            # 跳过输入结构，获取实际读取的数据
            memory_data = data[18:18+size]
            print(f"[+] 读取内存成功: PID={process_id}, 地址=0x{address:X}, 大小={size}")
            return memory_data
        else:
            print(f"[-] 读取内存失败，状态码: 0x{status:08X}")
            return None
    
    def write_process_memory(self, process_id, address, data):
        """写入进程内存"""
        # struct input {
        #     UINT64 processid;
        #     UINT64 startaddress;
        #     WORD bytestowrite;
        # }
        size = len(data)
        input_data = struct.pack("<QQH", process_id, address, size) + data
        status, _ = self.send_request(IOCTL_CE_WRITEMEMORY, input_data, 0)
        
        if status == 0:
            print(f"[+] 写入内存成功: PID={process_id}, 地址=0x{address:X}, 大小={size}")
            return True
        else:
            print(f"[-] 写入内存失败，状态码: 0x{status:08X}")
            return False
    
    def get_peprocess(self, process_id):
        """获取EPROCESS结构地址"""
        input_data = struct.pack("<I", process_id)
        status, data = self.send_request(IOCTL_CE_GETPEPROCESS, input_data, 8)
        
        if status == 0 and data:
            peprocess = struct.unpack("<Q", data)[0]
            print(f"[+] EPROCESS地址: 0x{peprocess:016X}")
            return peprocess
        else:
            print(f"[-] 获取EPROCESS失败，状态码: 0x{status:08X}")
            return None


def print_hex_dump(data, address=0):
    """打印十六进制数据"""
    for i in range(0, len(data), 16):
        hex_str = " ".join(f"{b:02X}" for b in data[i:i+16])
        ascii_str = "".join(chr(b) if 32 <= b < 127 else "." for b in data[i:i+16])
        print(f"  {address+i:08X}: {hex_str:<48} {ascii_str}")


def main():
    """主函数 - 测试示例"""
    print("=" * 60)
    print("DBK驱动Socket通信客户端测试")
    print("=" * 60)
    
    # 创建客户端
    client = DBKSocketClient()
    
    # 连接到驱动
    if not client.connect():
        return
    
    try:
        # 测试1: 获取驱动版本
        print("\n[测试1] 获取驱动版本")
        print("-" * 60)
        version = client.get_version()
        
        # 测试2: 打开当前进程
        print("\n[测试2] 打开当前进程")
        print("-" * 60)
        import os
        current_pid = os.getpid()
        print(f"当前进程PID: {current_pid}")
        handle = client.open_process(current_pid)
        
        # 测试3: 获取EPROCESS
        print("\n[测试3] 获取EPROCESS结构")
        print("-" * 60)
        peprocess = client.get_peprocess(current_pid)
        
        # 测试4: 读取内存（示例：读取进程自身的一些内存）
        print("\n[测试4] 读取进程内存")
        print("-" * 60)
        # 这里只是示例，实际地址需要根据具体情况调整
        test_address = 0x400000  # 典型的PE文件基址
        memory = client.read_process_memory(current_pid, test_address, 64)
        if memory:
            print("内存数据:")
            print_hex_dump(memory, test_address)
        
        print("\n" + "=" * 60)
        print("测试完成！")
        print("=" * 60)
        
    except KeyboardInterrupt:
        print("\n[!] 用户中断")
    except Exception as e:
        print(f"\n[-] 发生错误: {e}")
        import traceback
        traceback.print_exc()
    finally:
        # 断开连接
        client.disconnect()


if __name__ == "__main__":
    main()


